sources/api/schnauzer/src/v2/template.rs

```diff
@@ -0,0 +1,162 @@
+    #[serde(rename = "required-extensions")]
+    required_extensions: HashMap<ExtensionName, TemplateExtensionRequirements>,
+}
+
+/// Template extension requirements can be specified in two ways, similar to Cargo.toml:
+///   * extension = "version"
+///   * extension = { version = "version", helpers = ["helper1", "helper2"] }
+///
+/// The first form is simpler but cannot express a dependency on any helpers.
+#[derive(Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(untagged)]
+enum TemplateExtensionRequirements {
+    Version(ExtensionVersion),
+    VersionAndHelpers(DetailedTemplateExtensionRequirements),
+}
+
+/// Serialized structure of settings and handlebars helper requirements.
+#[derive(Deserialize, Debug, Clone, PartialEq, Eq, Default)]
+#[serde(deny_unknown_fields)]
+struct DetailedTemplateExtensionRequirements {
+    version: ExtensionVersion,
+    helpers: Option<Vec<HelperName>>,
+}
+
+impl TemplateFrontmatter {
+    /// Returns the name, version of each setting and the associated helpers required by this template.
+    pub fn extension_requirements(&self) -> impl Iterator<Item = ExtensionRequirement> + '_ {
+        self.required_extensions
+            .iter()
+            .map(|(extension_name, extension_requirements)| {
+                ExtensionRequirement::from_template_requirements(
+                    extension_name,
+                    extension_requirements,
+                )
+            })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Hash)]
+pub struct ExtensionRequirement {
+    pub name: ExtensionName,
+    pub version: ExtensionVersion,
+    pub helpers: Vec<HelperName>,
+}
+
+impl ExtensionRequirement {
+    fn from_template_requirements(
+        extension_name: &ExtensionName,
+        requirements: &TemplateExtensionRequirements,
+    ) -> Self {
+        match requirements {
+            TemplateExtensionRequirements::Version(version) => ExtensionRequirement {
+                name: extension_name.clone(),
+                version: version.clone(),
+                helpers: vec![],
+            },
+            TemplateExtensionRequirements::VersionAndHelpers(extension_requirements) => {
+                ExtensionRequirement {
+                    name: extension_name.clone(),
+                    version: extension_requirements.version.clone(),
+                    helpers: extension_requirements.helpers.clone().unwrap_or_default(),
+                }
+            }
+        }
+    }
+}
+
+impl FromStr for Template {
+    type Err = error::Error;
+
+    fn from_str(input_str: &str) -> Result<Template> {
+        // Potential improvement here would be to use `ariadne` crate to emit more helpful error messages
+        let mut parsed_pairs = TemplateParser::parse(Rule::template, input_str)
+            .context(error::GrammarParseSnafu)?
+            .flatten();
+
+        // If the template succesfully parses, we know there's always exactly one TOML document and one template body.
```

Comment:
Is this true even if front matter is not present? Is front matter always required in a template, or are there cases where front matter would be optional?

Reply:
As written, the frontmatter is required, but it can be an empty document. In these cases, the template is like this:

```
+++
template body
```

We actually have a template in Bottlerocket right now which is a static file, and looks just like this (I can't recall which offhand).

Reply:
I will add a testcase that does this.